### ISSUE TITLE
[XPU][oneDNN] Fused SDPA GPU page-fault / device lost when head_dim=64 with a rank-4 attn_mask spanning more than one tile on both axes (Sq>=17, Sk>=129); rank-2 mask is safe

### ISSUE THREAD

**Issue by 18856672421** [pytorch/pytorch#195319]
## Summary

On an Intel Arc (Alchemist-class) iGPU, the oneDNN fused SDPA kernel used by
`F.scaled_dot_product_attention` hard-crashes the device (GPU page fault →
device lost / process abort, **not** a catchable Python exception) **only when
all four** of the following hold:

1. `attn_mask` is a **rank-4** mask, shape `[B, 1, Sq, Sk]` (broadcast over heads);
2. `head_dim == 64`;
3. `Sq >= 17` (i.e. query axis spans a 2nd 16-wide tile);
4. `Sk >= 129` (i.e. key axis spans a 2nd 128-wide tile).

Changing **any single** condition makes it safe:

- `head_dim != 64` (e.g. 32 / 96 / 128) → OK;
- `Sq <= 16` **or** `Sk <= 128` (single tile on either axis) → OK;
- passing the **same mask reshaped to rank-2** `[Sq, Sk]` → OK;
- `attn_mask=None` → OK.

The crash is independent of batch size and of GQA: it reproduces at `B=1` and
with equal q/k/v head counts (no `enable_gqa`). It looks like an out-of-bounds
access in the fused kernel's second-tile handling, specific to head_dim=64,
that only triggers when a rank-4 mask is present.

This is related to (but distinct from) the known Alchemist head_dim=64 SDPA
problem in uxlfoundation/oneDNN#5510, which reports *wrong results* for d=64 on
Arc 770 / Arc 140T; here the failure mode is a hard device crash gated on the
rank-4 mask and the 16/128 tile boundaries.

## Environment

- PyTorch: **2.13.0+xpu** (official XPU wheel, Windows)
- Intel GPU driver: **32.0.101.8991**
- GPU: **Intel Arc Graphics 140T** (Alchemist / Xe-HPG class)
- OS: Windows 11
- dtype: `torch.bfloat16`
- Backend note on this build:
  - FLASH_ATTENTION: unavailable ("Torch XPU was not compiled with flash attention");
  - EFFICIENT_ATTENTION: prints "Memory Efficient Attention backend is not
    supported on XPU, falling back to math backend" → effectively MATH;
  - so the only fused path is the oneDNN fused kernel, which is the one that crashes.

## Minimal reproducer (crashes the process)

```python
import torch
import torch.nn.functional as F

dev, dt = "xpu", torch.bfloat16
B, H, Sq, Sk, D = 1, 14, 17, 129, 64          # <-- all four conditions met

q = torch.randn(B, H, Sq, D, device=dev, dtype=dt)
k = torch.randn(B, H, Sk, D, device=dev, dtype=dt)
v = torch.randn(B, H, Sk, D, device=dev, dtype=dt)
m = torch.ones(B, 1, Sq, Sk, device=dev, dtype=torch.bool)   # rank-4 broadcast mask

out = F.scaled_dot_product_attention(q, k, v, attn_mask=m)   # GPU page fault / device lost
torch.xpu.synchronize()
```

### Single-change controls (all OK — proves the four-way conjunction)

```python
# (a) different head_dim
F.scaled_dot_product_attention(q[..., :32], k[..., :32], v[..., :32], attn_mask=m)  # D=32  OK
# (b) query within one 16-tile
F.scaled_dot_product_attention(q[:, :, :16],  k, v, attn_mask=m[:, :, :16])         # Sq=16 OK
# (c) key within one 128-tile
F.scaled_dot_product_attention(q, k[:, :, :128], v[:, :, :128], attn_mask=m[..., :128])  # Sk=128 OK
# (d) identical mask, reshaped to rank-2
F.scaled_dot_product_attention(q, k, v, attn_mask=m.reshape(Sq, Sk))                # rank2 OK
# (e) no mask
F.scaled_dot_product_attention(q, k, v)                                             # OK
```

## Boundary sweep (run each config in a subprocess — a crash kills the worker)

```python
# probe.py  —  python probe.py            (driver, summarizes)
#             python probe.py w <D> <Sq> <Sk> <mask_rank>   (one isolated worker)
import sys, subprocess

def worker(D, Sq, Sk, mask_rank):
    import torch
    import torch.nn.functional as F
    dev, dt = "xpu", torch.bfloat16
    B, H = 1, 14
    q = torch.randn(B, H, Sq, D, device=dev, dtype=dt)
    k = torch.randn(B, H, Sk, D, device=dev, dtype=dt)
    v = torch.randn(B, H, Sk, D, device=dev, dtype=dt)
    if mask_rank == 4:
        m = torch.ones(B, 1, Sq, Sk, device=dev, dtype=torch.bool)
    else:
        m = torch.ones(Sq, Sk, device=dev, dtype=torch.bool)
    out = F.scaled_dot_product_attention(q, k, v, attn_mask=m)
    torch.xpu.synchronize()
    print("OK", tuple(out.shape), flush=True)

def probe(D, Sq, Sk, mask_rank):
    r = subprocess.run(
        [sys.executable, __file__, "w", str(D), str(Sq), str(Sk), str(mask_rank)],
        capture_output=True, text=True, timeout=120)
    low = (r.stdout + r.stderr).lower()
    if "ok" in low:
        return "OK"
    if any(s in low for s in ("device lost", "segmentation fault", "page fault",
                              "abort", "fault", "crash")):
        return "CRASH"
    return "ERR: " + (r.stderr.strip().splitlines() or ["?"])[-1][:80]

def main():
    grid = [
        # (D,  Sq,   Sk,   mask_rank)
        (64,  16,  128, 4),   # single tile both axes
        (64,  17,  128, 4),   # 2nd Q tile only
        (64,  16,  129, 4),   # 2nd K tile only
        (64,  17,  129, 4),   # 2nd tile on BOTH axes  -> crash
        (64,  33,  256, 4),   # further over tiles      -> crash
        (32,  17,  129, 4),   # same over-tile, hd!=64  -> OK
        (128, 17,  129, 4),   # same over-tile, hd!=64  -> OK
        (64,  17,  129, 2),   # same shape, rank-2 mask -> OK
        (64, 130, 1024, 4),   # realistic long context  -> crash
    ]
    for D, Sq, Sk, mr in grid:
        print(f"hd={D:<3} Sq={Sq:<4} Sk={Sk:<4} mask_rank={mr} -> {probe(D, Sq, Sk, mr)}",
              flush=True)

if __name__ == "__main__":
    if len(sys.argv) >= 6 and sys.argv[1] == "w":
        worker(int(sys.argv[2]), int(sys.argv[3]), int(sys.argv[4]), int(sys.argv[5]))
    else:
        main()
```

### Observed result

hd=64  Sq=16   Sk=128  mask_rank=4 -> OK
hd=64  Sq=17   Sk=128  mask_rank=4 -> OK
hd=64  Sq=16   Sk=129  mask_rank=4 -> OK
hd=64  Sq=17   Sk=129  mask_rank=4 -> ERR: level_zero backend failed with error: 20 (UR_RESULT_ERROR_DEVICE_LOST)
hd=64  Sq=33   Sk=256  mask_rank=4 -> ERR: level_zero backend failed with error: 20 (UR_RESULT_ERROR_DEVICE_LOST)
hd=32  Sq=17   Sk=129  mask_rank=4 -> OK
hd=128 Sq=17   Sk=129  mask_rank=4 -> OK
hd=64  Sq=17   Sk=129  mask_rank=2 -> OK
hd=64  Sq=130  Sk=1024 mask_rank=4 -> ERR: level_zero backend failed with error: 20 (UR_RESULT_ERROR_DEVICE_LOST)

=> crash requires **head_dim=64 AND rank-4 mask AND crossing both the 16-Q and
128-K tile boundaries**. Batch size (reproduces at B=1) and GQA (reproduces with
equal q/k/v heads) are not required.

## Real-world impact

Hit while running Qwen2-VL inference (transformers SDPA attention interface) via
MinerU on this GPU. In the actual model q has 14 heads vs 2 KV heads (GQA,
`enable_gqa=True`) and the padding mask is rank-4 `[B,1,Sq,Sk]`; once the
prefill context grows past the thresholds above (e.g. Sq=685, Sk=685, hd=64),
every attention layer crashes and takes down the whole process.

## Workaround (user-side, confirms the diagnosis)

Force the MATH backend

This is numerically correct and prevents the device crash, and it needs no
reshaping or per-batch splitting. However, MATH is a non-fused implementation
and is substantially slower than the native oneDNN fused kernel on this iGPU,
which noticeably slows down every affected attention layer during prefill.
That performance gap — having to stay on the slow MATH path for a shape the
fused kernel ought to support — is the reason for this report.

## Expected behavior

The fused oneDNN SDPA kernel should compute correct results for
head_dim=64 with a rank-4 mask that spans multiple 16/128 tiles (matching
CUDA and the rank-2-mask / other-head-dim behavior)

### Versions

Python version: 3.13.4 (tags/v3.13.4:8a526ec, Jun  3 2025, 17:46:04) [MSC v.1943 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-11-10.0.26200-SP0
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: True
XPU used to build PyTorch: 20260000
Intel GPU driver version:
* 32.0.101.8991 (20260824000000.******+***)
Intel GPU models onboard:
* Intel(R) Arc(TM) 140T GPU (16GB)
Intel GPU models detected:
* [0] _XpuDeviceProperties(name='Intel(R) Arc(TM) 140T GPU (16GB)', platform_name='Intel(R) oneAPI Unified Runtime over Level-Zero', type='gpu', device_id=0x7D51, driver_version='1.15.39183+3', total_memory=16831MB, local_mem_size=64KB, last_level_cache_size=8192KB, max_compute_units=128, memory_clock_rate=0MHz, memory_bus_width=64-bit, gpu_eu_count=128, gpu_subslice_count=16, max_work_group_size=1024, max_num_sub_groups=128, sub_group_sizes=[8 16 32], has_fp16=1, has_fp64=1, has_atomic64=1, is_integrated_gpu=1)
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False
Caching allocator config: N/A

CPU:
Name: Intel(R) Core(TM) Ultra 9 285H
Manufacturer: GenuineIntel
Family: 775
Architecture: 9
ProcessorType: 3
DeviceID: CPU0
CurrentClockSpeed: 2900
MaxClockSpeed: 2900
L2CacheSize: 28672
L2CacheSpeed: None
Revision: None

Versions of relevant libraries:
[pip3] dpcpp-cpp-rt==2026.0.0
[pip3] intel-cmplr-lib-rt==2026.0.0
[pip3] intel-cmplr-lib-ur==2026.0.0
[pip3] intel-cmplr-lic-rt==2026.0.0
[pip3] intel-opencl-rt==2026.0.0
[pip3] intel-openmp==2026.0.0
[pip3] intel-pti==0.17.0
[pip3] intel-sycl-rt==2026.0.0
[pip3] mkl==2026.0.0
[pip3] numpy==2.5.2
[pip3] onemkl-license==2026.0.0
[pip3] onemkl-sycl-blas==2026.0.0
[pip3] onemkl-sycl-dft==2026.0.0
[pip3] onemkl-sycl-lapack==2026.0.0
[pip3] onemkl-sycl-rng==2026.0.0
[pip3] onemkl-sycl-sparse==2026.0.0
[pip3] onnxruntime==1.29.0
[pip3] tbb==2023.0.0
[pip3] tcmlib==1.5.0
[pip3] torch==2.13.0+xpu
[pip3] torchaudio==2.11.0+xpu
[pip3] torchvision==0.28.0+xpu
[pip3] triton-xpu==3.7.2
[pip3] umf==1.1.0
[conda] Could not collect

cc @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10 @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @snadampal @EikanWang @fengyuan14 @guangyey @drisspg @liangel-02 @howardzhang-cv

**Comment by priyanshusky0** [comment]
Hi @18856672421 ,  
I came across this issue and would be interested in investigating it. The fact that the failure is isolated to head_dim=64 with a rank-4 mask and occurs specifically when both the Sq=16 and Sk=128 tile boundaries are crossed makes this particularly interesting.
I'd like to trace the XPU SDPA dispatch into the oneDNN fused implementation, identify the multi-tile/rank-4 mask path responsible for the invalid device access, and add regression coverage around the identified boundary conditions.
If you're not already working on a fix, would you be able to assign the issue to me? I'd be happy to investigate and work toward a PR.
Thanks!

**Comment by 18856672421** [comment]
Hi,@priyanshusky0 
Thanks for offering — please go ahead!

Attaching both dump files:

- `graph-104658….json`: real-model path, GQA rank-5, q=[7,2,7,454,64],
  external bf16 additive mask [7,1,1,454,454], Sq=Sk=454
- `graph-100002-8527608715665910638m.json`: minimal repro, rank-4,
  q=[1,14,17,64], external bf16 mask [1,1,17,129]

The crash occurs with an external mask in both rank-4 and GQA rank-5
forms, and consistently at `head_dim=64` when both the 16-Q and 128-K
tile boundaries are crossed.

For context, a contributor mentioned that this subgraph runs successfully
with oneDNN v3.13.2, whereas the current XPU PyTorch builds do not yet
bundle oneDNN 3.13.2 (the latest XPU nightly still ships 3.12.3).

On assignment: as the external reporter, I don't have triage/assign
permission, so I can't assign the issue myself. @malfet ,
could you please assign it to @priyanshusky0 ? In the meantime,
feel free to start from a fork — an assignment isn't required to open
a draft PR.

[graph-100002-8527608715665910638m.json](https://github.com/user-attachments/files/31637467/graph-100002-8527608715665910638m.json)

[graph-104658-5349023115352890067.json](https://github.com/user-attachments/files/31637474/graph-104658-5349023115352890067.json)

**Comment by guangyey** [comment]
@LuFinch could you please help take a look

**Comment by LuFinch** [comment]
LvTao help to check from OneDNN side and find v3.13.2 could help fix this issue, refer to https://github.com/uxlfoundation/oneDNN/issues/5933.

Hence this issue has dependency on OneDNN upgrade. We are doing acceptance test and will upgrade in coming Pytorch2.15.

**Comment by TaoLv** [comment]
@LuFinch  Please help to double check this case during the acceptance test as well. Thank you!

**Comment by LuFinch** [comment]
We don't have A770.. 

@18856672421 Could you help try with these wheel? 
Linux: https://github.com/pytorch/pytorch/actions/runs/33480437862?pr=195551
Windows: https://github.com/pytorch/pytorch/actions/runs/33480437954?pr=195551
It is Pytorch release2.14 + oneDNN v3.13. Let's see whether oneDNN v3.13 can help solve your issue.

Scroll the page to the bottom and you can download the wheel directly.

**Comment by 18856672421** [comment]
@LuFinch  Happy to help, but I should note I don't have an A770 — my GPU is Intel Arc 140T iGPU (Core Ultra 9 285H, Arrow Lake-H, Xe-LPG+). It's a supported XPU target in 2.14. Would a test result on this iGPU still be useful for verifying the oneDNN v3.13 change? If so I'll download the Windows wheel and report back.

**Comment by LuFinch** [comment]
@18856672421  I am not sure whether OneDNN v3.13 could help solve this issue on Arc 140T iGPU.  This issue seem to be platform specific. Our develop machines are PVC/B580/B70 and I can't reproduce this issue. Could you help verifying?

**Comment by 18856672421** [comment]
@LuFinch  Sure, happy to verify.

**Comment by 18856672421** [comment]
@LuFinch  Confirming the fix works on Arc 140T iGPU: both original repros and an end-to-end MinerU pipeline run cleanly and stably with Pytorch release2.14 + oneDNN v3.13.

**Comment by LuFinch** [comment]
@18856672421 Thanks for your help! Let's waiting for upgrade PR merge.

**Comment by 18856672421** [comment]
@LuFinch  You're welcome! Thanks for the quick fix and for providing the test wheels. I'll keep an eye on the upgrade PR and re-verify once it lands in a release. Feel free to ping me if any further Arc 140T / Arrow Lake-H testing is needed.

**Comment by TaoLv** [comment]
Thank you, @18856672421 and @LuFinch !
